lib/regex.h

```diff
@@ -0,0 +1,49 @@
+/*
+ * Cppcheck - A tool for static C/C++ code analysis
+ * Copyright (C) 2007-2023 Cppcheck team.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//---------------------------------------------------------------------------
+#ifndef regexH
+#define regexH
+//---------------------------------------------------------------------------
+
+#ifdef HAVE_RULES
+
+#include <functional>
+#include <memory>
+#include <string>
+
+class Regex
+{
+public:
+    explicit Regex(std::string pattern);
+
+    std::string compile();
+
+    using MatchFn = std::function<void(int start, int end)>;
+    std::string match(const std::string& str, const MatchFn& match)const ;
+
+private:
+    std::string mPattern;
+
+    class Data;
+    std::shared_ptr<Data> mData;
```

Comment:
This needs to be a shared pointer because the it is used in `Settings` are they are being copied (the copies in production code will hopefully go away soon) and we have dynamic memory.

This means that multiple threads will use this but as the used data is `const` there should not be any issues. I did not see anything related to thread safety in the PCRE documentation. I will check again though. I hope any potential issues will be caught by the CI.

Thinking about that we a `--rule` test which actually has multiple input files. Wil add that.